### DISCUSSION TITLE
Propper way to leverage `selectors` and `reactions` in `joinAdapters` for tightly coupled state slices

### DISCUSSION THREAD

**Discussion by pBouillon** [state-adapt/state-adapt#61] [Q&A] [answered]
Hi, I'm still exploring the usage of `StateAdapt` and its API and I tried to see how to take advantage of `joinAdapters` in a scenario where pieces of my state should react to the same action.

For that I tried to represent the score in a [Molkky game](https://www.molkky.com/wp/rules/)

My state consist of:
- The number of consecutive misses
- The number of points

I would like to add a `addFallenPins` method that:
- Increment the consecutive misses if no pin has been hit
- If at least one pin is hit and the misses streak was greater than 0, reset it
- Recompute the score based on the fallen pins

I've come to the following code but it seems a bit messy:

```ts
export const scoreAdapter = joinAdapters<Score>()({
  currentMissStreak: numberAdapter,
  points: molkkyScoreAdapter,
})(([selectors, reactions]) => ({
  addFallenPins: (state, fallenPins: Pin[]) => {
    // 👇 Add a miss if no pin is fallen
    const isMiss = fallenPins.length === 0;
    if (isMiss) {
      return {
        ...state,
        currentMissStreak: numberAdapter.increment(state.currentMissStreak),
      };
    }

    const wasOnMissStreak = state.currentMissStreak > 0;

    return {
      ...state,
      // 👇 Recompute the score
      points: molkkyScoreAdapter.addFallenPins(state.points, fallenPins),
      // 👇 Reset the streak if needed
      currentMissStreak: wasOnMissStreak
        ? numberAdapter.set(state.currentMissStreak, 0)
        : state.currentMissStreak,
    };
  },
}))();
```

I thought I could fan-out a reaction like: adding fallen pins would trigger a reaction on the streak, resetting it if needed, and it would also cause the score to be reevaluated based on the fallen pins.

I tried to dig look at the documentation and [a discussion](https://github.com/state-adapt/state-adapt/discussions/43) that looked similar, but it only references one slice of the state.

Did I miss something here, is there a better way to represent this logic?

**Comment by pBouillon** [comment]
**UPDATE:** I took a look at [one of your video](https://www.youtube.com/watch?v=g3FkpoOjFwc&list=PLA8RBi6wjQj_8EmpWz5A6m--A-MNbZXRd&index=5) where you are defining an adapter for a specific slice in order to avoid spreading the state.

I tried to do something similar, and it does look a bit more straightforward, but I'm still unsure this is the right place to do it in the `joinAdapters`:

```ts
export const missesStreakAdapter = createAdapter<number>()({
  update: (streak, fallenPins: Pin[]) =>
    fallenPins.length > 0 ? 0 : streak + 1,
});

export const scoreAdapter = joinAdapters<Score>()({
  currentMissesStreak: missesStreakAdapter,
  points: molkkyScoreAdapter,
})(([_selectors, _reactions]) => ({
  addFallenPins: (state, fallenPins: Pin[]) => ({
    ...state,
    points: molkkyScoreAdapter.addFallenPins(state.points, fallenPins),
    currentMissStreak: missesStreakAdapter.update(
      state.currentMissesStreak,
      fallenPins,
    ),
  }),
}))();
```

**Reply by mfp22** [reply, marked as answer]
I think this works, but you probably want grouped reactions. It's the 3rd possibility in the adapter builder syntax that joinAdapters also has: https://state-adapt.github.io/docs/core#buildadapter

**Reply by pBouillon** [reply]
It does sound like a better idea indeed, thanks a lot, I've come to this:

```ts
export const scoreAdapter = joinAdapters<Score>()({
  currentMissesStreak: missesStreakAdapter,
  points: molkkyScoreAdapter,
})({
  addFallenPins: {
    points: molkkyScoreAdapter.update,
    currentMissStreak: missesStreakAdapter.update,
  },
})();
```

And TS hints are indeed valid.

However I think that I didn't get the second option, using `selectors` and `reactions`. I guess the documentation is a bit too abstract for me and I don't get how to use them or their purpose then. Do you have any other resources or references to their usages?

**Reply by mfp22** [reply]
It's just more flexibility. Maybe it's more advanced. I had so many ideas of different syntaxes and settled on 3. Maybe the function should be the 3rd option in the docs because it is maybe more rare than grouped reactions. It's actually possible to define reactions that use selector memoization with the function syntax, but taking in a selectors cache object, which is the 4th argument of a reaction, and passing it into the 2nd argument of a selector function. And it's possible to define new reactions altogether. Like, maybe there are 2 properties that need to change, but they rely on the current state of a 3rd property to calculate their new states... Like I said, it's for flexibility, so I'm sure there are use cases. I don't have many examples because I myself haven't had to use that syntax.

**Reply by pBouillon** [reply]
I understand, it would be nice if some demo project could feature a use case of this syntax some day, to get a more practical example.
